tempesta_fw/http.c

```diff
@@ -4141,25 +4148,33 @@ tfw_http_msg_process(void *conn, TfwFsmData *data)
 {
 	int r = T_OK;
 	TfwStream *stream = &((TfwConn *)conn)->stream;
-	unsigned int trail = data->trail;
+	unsigned int data_left = data->len;
 	struct sk_buff *next;
 
 	if (data->skb->prev)
 		data->skb->prev->next = NULL;
-	for (next = data->skb->next; data->skb;
+	for (next = data->skb->next; data->skb && data_left > 0;
 	     data->skb = next, next = next ? next->next : NULL, data->off = 0)
 	{
 		if (likely(r == T_OK || r == T_POSTPONE)) {
 			data->skb->next = data->skb->prev = NULL;
-			data->trail = !next ? trail : 0;
+			data->len = min_t(unsigned int, data_left,
+					  data->skb->len - data->off);
+			data_left -= data->len;
 			r = TFW_CONN_H2(conn)
 				? tfw_h2_frame_process(conn, data)
 				: tfw_http_msg_process_generic(conn, stream, data);
 		} else {
-			kfree(data->skb);
+			kfree_skb(data->skb);
```

Comment:
Ar this stage `skb` is orphaned from queues and used only inside Templesta, in all surrounding functions we use `__kfree_skb()` without reference checking and tracing events. should it be `__kfree_skb()` also?

Reply:
Looks safe. Changed to `__kfree_skb()`.

Reply:
This code was extracted into #1349 and merged.